app/models/project.rb

```diff
@@ -145,29 +145,51 @@ def add_observation(obs)
     imgs.each { |img| images.push(img) }
   end
 
+  # Nimmo note: I remember this one was a lot of work to get right.
+  # IIRC, the Arel was definitely more efficient than the AR for this join.
+  # But it's been so long, i honestly don't remember. Viva la small PR!
+
   # Remove observation (and its images) from this project. Saves it.
   def remove_observation(obs)
     return unless observations.include?(obs)
 
     imgs = obs.images.select { |img| img.user_id == obs.user_id }
     if imgs.any?
-      img_ids = imgs.map(&:id).map(&:to_s).join(",")
       # Leave images which are attached to other observations
       # still attached to this project.
-      leave_these_img_ids = Image.connection.select_values(%(
-        SELECT io.image_id FROM images_observations io, observations_projects op
-        WHERE io.image_id IN (#{img_ids})
-          AND io.observation_id != #{obs.id}
-          AND io.observation_id = op.observation_id
-          AND op.project_id = #{id}
-      )).map(&:to_i)
+      select_ids = arel_select_leave_these_img_ids(obs, imgs)
+      # puts(select_ids.to_sql)
+      leave_these_img_ids = Image.connection.select_values(
+        select_ids.to_sql
+      ).map(&:to_i)
+      # puts(leave_these_img_ids)
       imgs.reject! { |img| leave_these_img_ids.include?(img.id) }
+      # puts(imgs.inspect)
     end
     observations.delete(obs)
     imgs.each { |img| images.delete(img) }
     update_attribute(:updated_at, Time.zone.now)
   end
 
+  # SELECT io.image_id FROM images_observations io, observations_projects op
+  # WHERE io.image_id IN (#{img_ids})
+  #   AND io.observation_id != #{obs.id}
+  #   AND io.observation_id = op.observation_id
+  #   AND op.project_id = #{id}
+  def arel_select_leave_these_img_ids(obs, imgs)
```

Comment:
Note:  I believe that's what's throwing the ABC rubocop is the innermost condition`.and(io[:observation_id].eq(op[:observation_id])`. 

It might seem like this condition is implied by the joins...through, and maybe AR would figure out to join them on `observation_id`. But Arel does not produce the correct SQL if I omit that inner condition.

Reply:
Disabled the cop here

Reply:
+1 on disabling the cop.

If it can be done easily (makes **my** head spin) consider writing Arel that returns the images to be deleted (rather than those to be retained) and re-writing the above method accordingly to (among other things) get rid of `reject`.

Reply:
Good idea, but makes my head spin too. I'll pass on the challenge.